spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpRequest.java

```diff
@@ -0,0 +1,146 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.http.client.reactive;
+
+import java.net.HttpCookie;
+import java.net.URI;
+import java.util.Collection;
+
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.reactive.client.ContentChunk;
+import org.eclipse.jetty.reactive.client.ReactiveRequest;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import org.springframework.core.io.buffer.DataBuffer;
+import org.springframework.core.io.buffer.DataBufferFactory;
+import org.springframework.core.io.buffer.DefaultDataBufferFactory;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.MediaType;
+import org.springframework.util.Assert;
+
+/**
+ * {@link ClientHttpRequest} implementation for the Jetty HTTP client.
+ *
+ * @author Sebastien Deleuze
+ * @since 5.1
+ */
+public class JettyClientHttpRequest extends AbstractClientHttpRequest {
+
+	private final Request jettyRequest;
+
+	private final DefaultDataBufferFactory bufferFactory = new DefaultDataBufferFactory();
+
+	private ReactiveRequest reactiveRequest;
+
+	private Mono<ClientHttpResponse> response;
+
+
+	public JettyClientHttpRequest(Request jettyRequest) {
+		this.jettyRequest = jettyRequest;
+	}
+
+
+	@Override
+	protected void applyHeaders() {
+		getHeaders().entrySet().forEach(e -> e.getValue().forEach(v -> this.jettyRequest.header(e.getKey(), v)));
+		if (!getHeaders().containsKey(HttpHeaders.ACCEPT)) {
+			this.jettyRequest.header(HttpHeaders.ACCEPT, "*/*");
+		}
+	}
+
+	@Override
+	protected void applyCookies() {
+		getCookies().values().stream().flatMap(Collection::stream)
+				.map(cookie -> new HttpCookie(cookie.getName(), cookie.getValue()))
+				.forEach(this.jettyRequest::cookie);
+	}
+
+	@Override
+	public HttpMethod getMethod() {
+		HttpMethod method = HttpMethod.resolve(this.jettyRequest.getMethod());
+		Assert.notNull(method, "Method must not be null");
+		return method;
+	}
+
+	@Override
+	public URI getURI() {
+		return this.jettyRequest.getURI();
+	}
+
+	@Override
+	public DataBufferFactory bufferFactory() {
+		return this.bufferFactory;
+	}
+
+	@Override
+	public Mono<Void> writeWith(Publisher<? extends DataBuffer> publisher) {
+		String contentType = jettyRequest.getHeaders().contains(HttpHeader.CONTENT_TYPE) ?
+				jettyRequest.getHeaders().getField(HttpHeader.CONTENT_TYPE).getValue():
+				MediaType.APPLICATION_OCTET_STREAM_VALUE;
+		Publisher<ContentChunk> chunks = Flux.from(publisher).map(DataBuffer::asByteBuffer).map(ContentChunk::new);
```

Comment:
Even though this class is using `DefaultDataBuffer`, which does not require releasing, I still think it would be a good idea to release them here anyway, especially if we make the data buffer factory configurable.

I think the release can be implemented as the `ContentChunk` callback (I am assuming that the callback would be called after the data has been written, not sure about that). I see that below you said that releasing in the callback does not work for the response, but perhaps it does for the request?